projects/capybara/htm/htm_network.py

```diff
@@ -0,0 +1,245 @@
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+# See the GNU Affero Public License for more details.
+#
+# You should have received a copy of the GNU Affero Public License
+# along with this program.  If not, see http://www.gnu.org/licenses.
+#
+# http://numenta.org/licenses/
+# ----------------------------------------------------------------------
+import numpy as np
+import htmsanity.nupic.runner as sanity
+
+from nupic.bindings.algorithms import SpatialPooler, TemporalMemory
+from nupic.encoders.random_distributed_scalar import (
+  RandomDistributedScalarEncoder)
+
+
+
+def _computeAnomalyScore(activeColumnsNZ,
+                         previouslyPredictiveCellsNZ,
+                         nCellsPerColumn):
+  """
+  The set of the region's active columns and the set of columns that have 
+  previously-predicted cells are used to calculate the anomaly score.
+  """
+
+  if type(previouslyPredictiveCellsNZ) != np.ndarray:
+    previouslyPredictiveCellsNZ = np.array(previouslyPredictiveCellsNZ)
+
+  previouslyPredictiveColumnsNZ = previouslyPredictiveCellsNZ / nCellsPerColumn
+  burstingColumnsNZ = np.setdiff1d(activeColumnsNZ,
+                                   previouslyPredictiveColumnsNZ)
+  return len(burstingColumnsNZ) / float(len(activeColumnsNZ))
+
+
+
+def _computePredictedActiveCells(activeCellsNZ,
+                                 previouslyPredictiveCellsNZ):
+  predictedActive = np.intersect1d(activeCellsNZ, previouslyPredictiveCellsNZ)
+  assert len(predictedActive) <= activeCellsNZ
+  return predictedActive
+
+
+
+class BaseNetwork(object):
+  def __init__(self, inputMin=None, inputMax=None, runSanity=False):
+
+    self.inputMin = inputMin
+    self.inputMax = inputMax
+    self.runSanity = runSanity
+
+    self.encoder = None
+    self.encoderOutput = None
+    self.sp = None
+    self.spOutput = None
+    self.spOutputNZ = None
+    self.tm = None
+    self.anomalyScore = None
+    if runSanity:
+      self.sanity = None
+
+    self.defaultEncoderResolution = 0.0001
+    self.numColumns = 2048
+    self.cellsPerColumn = 32
+
+    self.predictedActiveCells = None
+    self.previouslyPredictiveCells = None
+
+
+  def initialize(self):
+
+    # Scalar Encoder
+    resolution = self.getEncoderResolution()
+    self.encoder = RandomDistributedScalarEncoder(resolution, seed=42)
+    self.encoderOutput = np.zeros(self.encoder.getWidth(), dtype=np.uint32)
+
+    # Spatial Pooler
+    spInputWidth = self.encoder.getWidth()
+    self.spParams = {
+      "globalInhibition": True,
+      "columnDimensions": [self.numColumns],
+      "inputDimensions": [spInputWidth],
+      "potentialRadius": spInputWidth,
+      "numActiveColumnsPerInhArea": 40,
+      "seed": 1956,
+      "potentialPct": 0.8,
+      "boostStrength": 0.0,
+      "synPermActiveInc": 0.003,
+      "synPermConnected": 0.2,
```

Comment:
If you just want to test the value of TM, you may consider turning SP learning off by setting `synPermActiveInc` and `synPermInactiveDec` to zero.